lib/Dialect/AIE/Transforms/AIEAssignBufferDescriptorIDs.cpp

```diff
@@ -0,0 +1,174 @@
+  void runOnOperation() override {
+    DeviceOp targetOp = getOperation();
+    const AIETargetModel &targetModel = targetOp.getTargetModel();
+
+    auto memOps = llvm::to_vector_of<TileElement>(targetOp.getOps<MemOp>());
+    llvm::append_range(memOps, targetOp.getOps<MemTileDMAOp>());
+    llvm::append_range(memOps, targetOp.getOps<ShimDMAOp>());
+    for (TileElement memOp : memOps) {
+      int col = memOp.getTileID().col;
+      int row = memOp.getTileID().row;
+      BdIdGenerator gen(col, row, targetModel);
+
+      auto dmaOps = llvm::to_vector_of<DMAOp>(
+          memOp.getOperation()->getRegion(0).getOps<DMAOp>());
+      if (!dmaOps.empty()) {
+        for (auto dmaOp : dmaOps) {
+          auto bdRegions = dmaOp.getBds();
+          for (auto &bdRegion : bdRegions) {
+            auto &block = bdRegion.getBlocks().front();
+            DMABDOp bd = *block.getOps<DMABDOp>().begin();
+            if (bd.getBdId().has_value())
+              gen.assignBdId(bd.getBdId().value());
+            else
+              bd.setBdId(gen.nextBdId(dmaOp.getChannelIndex()));
+          }
+        }
+      } else {
+        DenseMap<Block *, int> blockChannelMap;
+        // Associate with each block the channel index specified by the
+        // dma_start
+        for (Block &block : memOp.getOperation()->getRegion(0))
+          for (auto op : block.getOps<DMAStartOp>()) {
+            int chNum = op.getChannelIndex();
+            blockChannelMap[&block] = chNum;
+            Block *dest = op.getDest();
+            while (dest) {
+              blockChannelMap[dest] = chNum;
+              if (dest->hasNoSuccessors())
+                break;
+              dest = dest->getSuccessors()[0];
+              if (blockChannelMap.contains(dest))
+                dest = nullptr;
+            }
+          }
+
+        for (Block &block : memOp.getOperation()->getRegion(0)) {
+          if (block.getOps<DMABDOp>().empty())
+            continue;
+          assert(blockChannelMap.count(&block));
+          DMABDOp bd = (*block.getOps<DMABDOp>().begin());
+          if (bd.getBdId().has_value())
+            gen.assignBdId(bd.getBdId().value());
+          else
+            bd.setBdId(gen.nextBdId(blockChannelMap[&block]));
+        }
+      }
+    }
+    for (TileElement memOp : memOps) {
+      auto dmaOps = llvm::to_vector_of<DMAOp>(
+          memOp.getOperation()->getRegion(0).getOps<DMAOp>());
+      if (!dmaOps.empty())
+        for (auto dmaOp : dmaOps) {
+          auto bdRegions = dmaOp.getBds();
+          for (auto *bdRegionIt = bdRegions.begin();
+               bdRegionIt != bdRegions.end();) {
+            auto &block = bdRegionIt->getBlocks().front();
+            DMABDOp bd = *block.getOps<DMABDOp>().begin();
+            std::optional<int> nextBdId;
+            if (++bdRegionIt != bdRegions.end())
+              nextBdId =
+                  (*bdRegionIt->getBlocks().front().getOps<DMABDOp>().begin())
```

Comment:
*bd?

Reply:
Dunno what you're suggesting? Rename `bdRegionIter` to `bd` or I should be derefing one of the floating/existing `bd` assignments here?

Reply:
The expressions here are large and somewhat redundant..  I think line 124/125 already implements most of the expression here.  Are there ways of simplifying the expressions?  Maybe factor out a lambda?